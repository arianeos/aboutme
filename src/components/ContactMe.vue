<template>
  <div class="white--text indigo lighten-3 mt-12">
    <v-container class="pa-12">
      <Alert :type="type" :message="alertMessage" v-show="alertMessage" />
      <h2 align="center" class="pb-6 text-h2">CONTACT ME</h2>
      <v-form ref="form">
        <v-text-field
          v-model="contact.name"
          id="name"
          label="*Name"
          :rules="requiredRules"
          required
          solo
          filled
          background-color="white"
          color="white"
        ></v-text-field>
        <v-text-field
          v-model="contact.email"
          id="email"
          label="*Email"
          :rules="emailRules"
          required
          solo
          filled
          background-color="white"
          color="white"
        ></v-text-field>
        <v-text-field
          v-model="contact.subject"
          id="subject"
          label="Subject"
          solo
          filled
          background-color="white"
          color="white"
        ></v-text-field>
        <v-textarea
          v-model="contact.message"
          label="*Message"
          id="message"
          :rules="requiredRules"
          required
          solo
          filled
          background-color="white"
        ></v-textarea>
        <v-btn right class="mr-4" @click="send">
          Send
        </v-btn>
      </v-form>
    </v-container>
  </div>
</template>
<script lang="ts">
import emailjs from "emailjs-com";
import Vue from "vue";
import Alert from "../components/Alert.vue";

export default Vue.extend({
  name: "ContactMe",
  components: {
    Alert
  },
  data() {
    return {
      contact: {
        name: "",
        email: "",
        subject: "",
        message: ""
      },
      requiredRules: [(v: any) => !!v || "This field is required"],
      emailRules: [
        (v: any) => !!v || "E-mail is required",
        (v: any) => /.+@.+/.test(v) || "E-mail must be valid"
      ],
      alertMessage: "",
      type: "success"
    };
  },
  computed: {
    isFormValid(): boolean {
      return (this.$refs.form as Vue & { validate: () => boolean }).validate();
    }
  },
  methods: {
    send(): void {
      if (!this.isFormValid) {
        return;
      }

      emailjs
        .send(
          process.env.VUE_APP_SERVICE_ID,
          process.env.VUE_APP_TEMPLATE_ID,
          this.contact,
          process.env.VUE_APP_USER_ID
        )
        .then(_ => {
          this.alertMessage = "Email successfully sent";
          this.cleanAlertMessage();
        })
        .catch(error => {
          console.error("FAILED...", error);
        });
    },
    cleanAlertMessage() {
      setTimeout(() => {
        this.alertMessage = "";
      }, 10000);
    }
  }
});
</script>
