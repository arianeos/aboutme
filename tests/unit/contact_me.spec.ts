import { shallowMount, mount, createLocalVue } from "@vue/test-utils";
import ContactMe from "@/components/ContactMe.vue";
import Vuetify from "vuetify";
import Vue from "vue";
import emailjs from "emailjs-com";

const localVue = createLocalVue();

Vue.use(Vuetify);

describe("ContactMe", () => {
  it("should renders the component and matches snapshot", () => {
    const wrapper = shallowMount(ContactMe, { localVue });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("when form is invalid, should not send an email when clicks on send btn", async () => {
    const wrapper = mount(ContactMe, {
      localVue
    });
    jest.spyOn(emailjs, "send");
    wrapper.find("button").trigger("click");
    expect(emailjs.send).not.toHaveBeenCalled();
  });

  it("when form is valid, should send an email when clicks on send btn", () => {
    const wrapper = mount(ContactMe, { localVue });
    jest.spyOn(emailjs, "send");
    wrapper.find("#name").setValue("Test One");
    wrapper.find("#email").setValue("test_one@test.com");
    wrapper.find("#subject").setValue("Subject of the email");
    wrapper.find("#message").setValue("Message of the email");
    wrapper.find("button").trigger("click");
    expect(emailjs.send).toHaveBeenCalled();
  });
});
