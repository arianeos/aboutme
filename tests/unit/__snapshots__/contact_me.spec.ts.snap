// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactMe should renders the component and matches snapshot 1`] = `
<div class="white--text indigo lighten-3 mt-12">
  <v-container-stub tag="div" class="pa-12">
    <alert-stub message="" type="success" style="display: none;"></alert-stub>
    <h2 align="center" class="pb-6 text-h2">CONTACT ME</h2>
    <v-form-stub>
      <v-text-field-stub color="white" errorcount="1" errormessages="" messages="" rules="v => !!v || &quot;This field is required&quot;" successmessages="" value="" backgroundcolor="white" id="name" label="*Name" loaderheight="2" clearicon="$clear" filled="true" solo="true" type="text" required=""></v-text-field-stub>
      <v-text-field-stub color="white" errorcount="1" errormessages="" messages="" rules="v => !!v || &quot;E-mail is required&quot;,v => /.+@.+/.test(v) || &quot;E-mail must be valid&quot;" successmessages="" value="" backgroundcolor="white" id="email" label="*Email" loaderheight="2" clearicon="$clear" filled="true" solo="true" type="text" required=""></v-text-field-stub>
      <v-text-field-stub color="white" errorcount="1" errormessages="" messages="" rules="" successmessages="" value="" backgroundcolor="white" id="subject" label="Subject" loaderheight="2" clearicon="$clear" filled="true" solo="true" type="text"></v-text-field-stub>
      <v-textarea-stub errorcount="1" errormessages="" messages="" rules="v => !!v || &quot;This field is required&quot;" successmessages="" value="" backgroundcolor="white" id="message" label="*Message" loaderheight="2" clearicon="$clear" filled="true" solo="true" type="text" rowheight="24" rows="5" required=""></v-textarea-stub>
      <v-btn-stub tag="button" activeclass="" right="true" type="button" class="mr-4">
        Send
      </v-btn-stub>
    </v-form-stub>
  </v-container-stub>
</div>
`;
